src/unix/poll.c

```diff
@@ -550,54 +646,34 @@ sysreturn epoll_ctl(int epfd, int op, int fd, struct epoll_event *event)
         rv = remove_fd(e, fd);
         break;
     case EPOLL_CTL_MOD:
-	epoll_debug("   modifying %d, events 0x%x, data 0x%lx\n", fd, event->events, event->data);
+        epoll_debug("   modifying %d, events 0x%x, data 0x%lx\n", fd, event->events, event->data);
         rv = remove_fd(e, fd);
         if (rv == 0)
             rv = epoll_add_fd(e, fd, event->events, event->data);
         break;
     default:
-	msg_err("unknown op %d\n", op);
-	rv = -EINVAL;
+        msg_err("unknown op %d\n", op);
+        rv = -EINVAL;
     }
+    spin_unlock(&e->fds_lock);
 
     fdesc_put(&e->f);
     return rv;
 }
 
 /* XXX build these out */
-#define POLLFDMASK_READ		(EPOLLIN | EPOLLHUP | EPOLLERR)
-#define POLLFDMASK_WRITE	(EPOLLOUT | EPOLLHUP | EPOLLERR)
-#define POLLFDMASK_EXCEPT	(EPOLLPRI)
+#define POLLFDMASK_READ     (EPOLLIN | EPOLLHUP | EPOLLERR)
+#define POLLFDMASK_WRITE    (EPOLLOUT | EPOLLHUP | EPOLLERR)
+#define POLLFDMASK_EXCEPT   (EPOLLPRI)
 
-closure_function(1, 2, void, select_notify,
-                 epollfd, efd,
-                 u64, notify_events,
-                 thread, t)
+static inline void select_notify(epollfd efd, epoll_blocked w, u64 events)
 {
-    epollfd efd = bound(efd);
-    list l = list_get_next(&efd->e->blocked_head);
-
-    if (notify_events == NOTIFY_EVENTS_RELEASE) {
-        epoll_debug("efd->fd %d unregistered\n", efd->fd);
-        efd->registered = false;
-        closure_finish();
+    spin_lock(&w->lock);
```

Comment:
Since w->nfds is set at the time the syscall is made and remains the same for the duration, I didn't bother with a copy and just moved this check outside the lock.